### PR TITLE
Performance fixes

### PR DESCRIPTION
**Pull request by jackgerrits**
These performance fixes are critical for the next step in the label/prediction pr.

- Change to using references with `v_array` where shallow copies were being used
- Add new print functions that use `v_array` references and deprecate the old by value ones
- Use `std::sort` instead of `qsort` in `features::sort`. Is able to make use of inlining because of the template, giving better runtime performance.

### CODE REVIEW

File: java/src/main/c++/vowpalWabbit_learner_VWMultilabelsLearner.cc [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 jobject multilabel_predictor(example *vec, JNIEnv *env)
 {
-  v_array<uint32_t> labels = vec->pred.multilabels.label_v;
+  v_array<uint32_t>& labels = vec->pred.multilabels.label_v;
```
Comment (peterychang) [inline java/src/main/c++/vowpalWabbit_learner_VWMultilabelsLearner.cc]:
```suggestion
  auto& labels = vec->pred.multilabels.label_v;
```

File: java/src/main/c++/vowpalWabbit_learner_VWScalarsLearner.cc [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 jfloatArray scalars_predictor(example *vec, JNIEnv *env)
 {
-  v_array<float> scalars = vec->pred.scalars;
+  v_array<float>& scalars = vec->pred.scalars;
```
Comment (peterychang) [inline java/src/main/c++/vowpalWabbit_learner_VWScalarsLearner.cc]:
```suggestion
  auto& scalars = vec->pred.scalars;
```

File: vowpalwabbit/cb_adf.cc [hunk 3]
```diff
@@ -97,8 +97,7 @@ struct cb_adf
 
 CB::cb_class get_observed_cost(multi_ex& examples)
 {
-  CB::label ld;
-  ld.costs = v_init<cb_class>();
+  CB::label* ld;
```
Comment (peterychang) [inline vowpalwabbit/cb_adf.cc]:
```suggestion
  CB::label* ld = nullptr;
```

File: vowpalwabbit/global_data.cc [hunk 4]
```diff
@@ -122,7 +138,25 @@ void print_raw_text(int f, std::string s, v_array<char> tag)
...
+  ssize_t len = ss.str().size();
+  ssize_t t = io_buf::write_file_or_socket(f, ss.str().c_str(), (unsigned int)len);
+  if (t != len)
+  {
+    std::cerr << "write error: " << strerror(errno) << std::endl;
+  }
+}
+
+
+void print_raw_text_by_ref(int f, std::string s, const v_array<char>& tag)
```
Comment (peterychang) [inline vowpalwabbit/global_data.cc]:
```suggestion
void print_raw_text_by_ref(int f, const std::string& s, const v_array<char>& tag)
```

File: vowpalwabbit/kernel_svm.cc [hunk 5]
```diff
@@ -439,7 +439,7 @@ float kernel_function(const flat_example* fec1, const flat_example* fec2, void*
   return 0;
 }
 
-float dense_dot(float* v1, v_array<float> v2, size_t n)
+float dense_dot(float* v1, v_array<float>& v2, size_t n)
```
Comment (peterychang) [inline vowpalwabbit/kernel_svm.cc]:
```suggestion
float dense_dot(float* v1, const v_array<float>& v2, size_t n)
```

File: vowpalwabbit/cb_explore_adf_common.h [hunk 6]
```diff
@@ -164,15 +164,15 @@ void cb_explore_adf_base<ExploreType>::output_example(vw& all, multi_ex& ec_seq)
   {
     std::string outputString;
     std::stringstream outputStringStream(outputString);
-    v_array<CB::cb_class> costs = ec.l.cb.costs;
+    v_array<CB::cb_class>& costs = ec.l.cb.costs;
```
Comment (peterychang) [inline vowpalwabbit/cb_explore_adf_common.h]:
```suggestion
    const auto& costs = ec.l.cb.costs;
```

File: vowpalwabbit/feature_group.h [hunk 7]
```diff
@@ -346,36 +348,44 @@ struct features
...
     if (!space_names.empty())
     {
-      v_array<feature_slice> slice = v_init<feature_slice>();
+      std::vector<feature_slice> slice;
+      slice.reserve(indicies.size());
       for (size_t i = 0; i < indicies.size(); i++)
       {
-        feature_slice temp = {values[i], indicies[i] & parse_mask, *space_names[i].get()};
-        slice.push_back(temp);
+        slice.push_back({values[i], indicies[i] & parse_mask, *space_names[i].get()});
```
Comment (peterychang) [inline vowpalwabbit/feature_group.h]: nitpick: Its a lot better to have a constructor and use emplace_back

Comment (jackgerrits) [reply vowpalwabbit/feature_group.h]: I agree, but that's out of scope for now

File: vowpalwabbit/feature_group.h [hunk 8]
```diff
@@ -346,36 +348,44 @@ struct features
...
     {
-      v_array<feature> slice = v_init<feature>();
+      std::vector<feature> slice;
+      slice.reserve(indicies.size());
+
       for (size_t i = 0; i < indicies.size(); i++)
       {
-        feature temp = {values[i], indicies[i] & parse_mask};
-        slice.push_back(temp);
+        slice.push_back({values[i], indicies[i] & parse_mask});
```
Comment (peterychang) [inline vowpalwabbit/feature_group.h]: same as above

**Pull request merged**